cosmos_core/src/logic/logic_graph.rs

```diff
@@ -0,0 +1,797 @@
+                    .groups
+                    .get_mut(&group_id)
+                    .expect("Removed logic port's group should have a vector of ports.");
+                // Delete it from the set of ports of its group.
+                match port_type {
+                    PortType::Input => drop(logic_group.consumers.remove(&port)),
+                    PortType::Output => drop(logic_group.producers.remove(&port)),
+                };
+
+                // Ping all inputs in this group to let them know this output port is gone.
+                if port_type == PortType::Output {
+                    for &input_port in self.groups.get(&group_id).expect("Port should have logic group.").consumers.iter() {
+                        evw_logic_input.send(LogicInputEvent {
+                            block: StructureBlock::new(input_port.coords),
+                            entity: structure.get_entity().expect("Structure should have entity."),
+                        });
+                    }
+                }
+            }
+
+            // Delete the port.
+            match port_type {
+                PortType::Input => &mut self.input_port_group_id,
+                PortType::Output => &mut self.output_port_group_id,
+            }
+            .remove(&port);
+        }
+    }
+
+    /// Adds a logic block, along with all of its ports and wire connections, to the graph.
+    /// If the added block has wire connections, merges adjacent [`LogicGroup`]s into a single group.
+    pub fn add_logic_block(
+        &mut self,
+        logic_block: &LogicBlock,
+        coords: BlockCoordinate,
+        structure: &Structure,
+        blocks: &Registry<Block>,
+        logic_blocks: &Registry<LogicBlock>,
+        evw_logic_output: &mut EventWriter<LogicOutputEvent>,
+        evw_logic_input: &mut EventWriter<LogicInputEvent>,
+    ) {
+        // Adding input faces as consumers to their connected group, or a new group if there is no connected group.
+        for input_face in logic_block.input_faces() {
+            self.placed_port(
+                coords,
+                input_face,
+                PortType::Input,
+                structure,
+                blocks,
+                logic_blocks,
+                evw_logic_output,
+                evw_logic_input,
+            )
+        }
+
+        // Adding output faces as consumers to their connected group, or a new group if there is no connected group.
+        for output_face in logic_block.output_faces() {
+            self.placed_port(
+                coords,
+                output_face,
+                PortType::Output,
+                structure,
+                blocks,
+                logic_blocks,
+                evw_logic_output,
+                evw_logic_input,
+            )
+        }
+
+        // Connect wire faces to all existing groups (by creating one new group that includes all adjacent groups).
+        if logic_block.wire_faces().count() > 0 {
```

Comment:
does this have the .is_empty() method? also early return here would be ideal for less indentation